### DISCUSSION TITLE
Understanding static vs dynamic routes prefetching

### DISCUSSION THREAD

**Discussion by gopinav** [vercel/next.js#51488] [App Router]
In the docs, there is a section on [prefetching](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#prefetching). Although I understood the concept of prefetching, I did not understand the difference between static and dynamic routes prefetching. I tried to inspect the network requests and understand what exactly is the difference but was unable to. Can someone help break this down with an example, please? Here's a dynamic route setup I've been working with.

![image](https://github.com/vercel/next.js/assets/13146172/5984dd2b-b1eb-450a-b8a1-266e94624614)

(From docs)
### Static and Dynamic Routes:

If the route is static, all the Server Component payloads for the route segments will be prefetched.

If the route is dynamic, the payload from the first shared layout down until the first loading.js file is prefetched. This reduces the cost of prefetching the whole route dynamically and allows [instant loading states](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming#instant-loading-states) for dynamic routes.

I especially could not understand this line: "the payload from the first shared layout down until the first loading.js file is prefetched"

**Comment by icyJoseph** [comment]
Hi,

Maybe if we think of GetServerSideProps and GetStatisProps, and then move into App Dir, we can understand this better.

See, in Pages Dir there are two things that drive a page, the JavaScript and its initial JSON, also called pageProps. 

When a page is pre-rendered, the initial JSON for it, is generated too. Pages that are rendered on demand do not have this. 

After build time, all pages, have their JS ready to go.

> This is because we needed to do something called Hydration, client side, so we needed to replay the server side render pass, by using the same inputs (JSON props), and collecting the same outputs (otherwise hydration errors occur), but also because when doing SPA navigation into a page, we do full 100% CSR.

So now, prefetching means:

- For a pre-rendered page: Get the JavaScript and initial JSON 
- For an on-demand rendered page: Get the JavaScript, but do not get the JSON because we need this on-demand

When the user clicks the link, or navigates, to this page, if it was pre-rendered, you have everything you need to render it. If it is on-demand page, we wait for the JSON to return, as promised, it is the latest available at the time the user navigated, and once that comes back, we render. We only wait for one thing.

This is of course a simplified model. 

In the App Dir, it is somewhat similar, static pages are flushed into HTML. In some cases, there might not even be JavaScript because it is only needed for Client Components. So to prefetch such page, to prepare for SPA navigation into the page, you just need the React Server Components props (.rsc file), to render the new tree, without technically having to do hydration. 

If it is a dynamic page, you can only prefetch up to a non-dynamic part of the tree, which hopefully is a loading fallback, then as the user does SPA navigation into that, we can use the prefetched `.rsc`, to render the new tree, while the data is fetched server side and a complementary `.rsc` is returned, which tells React what to use to patch the loading state, with actual data.

Simplified model too, but maybe it answer your question?

**Reply by gopinav** [reply]
Hi Joesph, thank you for taking the time to reply. I have a few follow-up questions:

> If it is a dynamic page, you can only prefetch up to a non-dynamic part of the tree, which hopefully is a loading fallback, then as the user does SPA navigation into that, we can use the prefetched .rsc, to render the new tree

Completely makes sense. However, in my products `page.tsx`, with three links on the page to `/products/1`, `/products/2`, `/products/3`, when I hover over each of the three links, I see prefetching in the network tab. If we are prefetching and caching the non-dynamic result for `/products/1`, shouldn't Next.js skip network requests for product Id 2 and 3?

I'm not able to make sense of the network requests:
<img width="1823" alt="image" src="https://github.com/vercel/next.js/assets/13146172/1c6c6035-e467-42f8-9b5f-e854e2b7b0f0">

Also, is the below what you refer to as rsc props? (Just for my understanding)
<img width="1735" alt="image" src="https://github.com/vercel/next.js/assets/13146172/7d8c3e9c-1ee4-4603-9cca-28370b02d0b5">

Thanks again!

**Reply by icyJoseph** [reply]
Yeah I'd say this is where some of the simplified model concepts break down for me. I had a similar issue with a page that used the cookies function, which means that I want it to be rendered strictly on demand, but you know what, it was prefetched... so I always had stale data on it. 

I solved it by using a server action, with `revalidatePath`.

In this app, when you throw a Pokeball, regardless of the outcome, the Pokemon is listed in collections. It works flawlessly in Pages Dir with the almighty GetServerSideProps. In app Dir, not so much, neither by using cookies or `force-dynamic`, it is always prefetched. I think it is a bug, but maybe I am just being dense. Anyway, server actions fixed it.

- Broken:
   - Demo: https://next-js-react-ct9wkgsaq-icyjoseph.vercel.app/pokemon/capture 
   - Code up to that point: https://github.com/icyJoseph/next-js-react-gbg/tree/9b955e66c758bf7ab3be8cb7f5517f713738b94a/demo-app
- Fixed:
  - Demo: https://next-js-react-gbg-git-feat-app-router-rsc-icyjoseph.vercel.app/pokemon/capture
  - Code: https://github.com/icyJoseph/next-js-react-gbg/tree/feat/app-router-rsc/demo-app

And the pages dir implementation https://next-js-react-gbg.vercel.app/

Related issue: https://github.com/vercel/next.js/issues/49865

So yeah, there's some stuff that doesn't add up 🤷

**Reply by gopinav** [reply]
Yeah, would love to see some more docs on this. Thanks again, Joesph!
